google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -0,0 +1,199 @@
+        return self._match_rules
+
+    @property
+    def verify_rules(self):
+        """The FirewallRules used to check policies.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        if not self._verify_rules:
+            self._verify_rules = self.create_rules(self._verify_policies)
+        return self._verify_rules
+
+    def find_policy_violations(self, firewall_policies):
+        """Finds policy violations in a list of firewall policies.
+
+        Args:
+          firewall_policies (list): A list of FirewallRule.
+
+        Yields:
+          iterable: A generator of RuleViolations.
+        """
+        if self.mode == scanner_rules.RuleMode.MATCHES:
+            if (len(firewall_policies) != len(self.match_rules) or
+                    any([is_rule_exists_violation(rule, firewall_policies)
+                         for rule in self.match_rules])):
+                yield self._create_violation(
+                    firewall_policies, 'FIREWALL_MATCHES_VIOLATION')
+        if self.mode == scanner_rules.RuleMode.REQUIRED:
+            if any([is_rule_exists_violation(rule, firewall_policies)
+                    for rule in self.match_rules]):
+                yield self._create_violation(
+                    firewall_policies, 'FIREWALL_REQUIRED_VIOLATION')
+        for policy in firewall_policies:
+            if not any([policy > rule for rule in self.match_rules]):
+                continue
+            if self.mode == scanner_rules.RuleMode.WHITELIST:
+                if is_whitelist_violation(self.verify_rules, policy):
+                    yield self._create_violation(
+                        [policy], 'FIREWALL_WHITELIST_VIOLATION')
+            if self.mode == scanner_rules.RuleMode.BLACKLIST:
+                if is_blacklist_violation(self.verify_rules, policy):
+                    yield self._create_violation(
+                        [policy], 'FIREWALL_BLACKLIST_VIOLATION')
+
+    def _create_violation(self, policies, violation_type):
+        """Creates a RuleViolation.
+
+        Args:
+          policies (list): A list of FirewallRule that violate the policy.
+          violation_type (str): The type of violation.
+
+        Returns:
+          RuleViolation: A RuleViolation for the given policies.
+
+        Raises:
+          ValueError: If no policies are passed in.
+        """
+        if not policies:
+            raise ValueError('No policies in violation')
+        return self.RuleViolation(
+            resource_type='firewall_policy',
+            resource_id=policies[0].project_id,
+            rule_name=self.name,
+            violation_type=violation_type,
+            policy_names=[p.name for p in policies])
+
+
+    # Rule violation.
+    # resource_type: string
+    # resource_id: string
+    # rule_name: string
+    # rule_index: int
+    # violation_type: FIREWALL_VIOLATION
+    # policy_name: string
+    RuleViolation = namedtuple('RuleViolation',
```

Comment:
Is there a reason to nest this in the class? I think it could be a module level class.

Reply:
Done